tpm2/policy.go

```diff
@@ -178,37 +180,92 @@ func createPcrPolicyCounterImpl(tpm *tpm2.TPMContext, handle tpm2.Handle, update
 	public.Attrs |= tpm2.AttrNVWritten
 
-	value, err := tpm.NVReadCounter(index, index, hmacSession)
+	// Read the current value
+	value, err = tpm.NVReadCounter(index, index, hmacSession)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	succeeded = true
 	return public, value, nil
 }
 
-// createPcrPolicyCounter creates and initializes a NV counter that is associated with a sealed key object
+// ensurePcrPolicyCounter creates and initializes a NV counter that is associated with a sealed key object
 // and is used for implementing PCR policy revocation.
 //
 // The NV index will be created with attributes that allow anyone to read the index, and an authorization
-// policy that permits TPM2_NV_Increment with a signed authorization policy. The caller must ensure that the
-// updateKey argument is a valid public key.
-var createPcrPolicyCounter = func(tpm *tpm2.TPMContext, handle tpm2.Handle, updateKey *tpm2.Public, hmacSession tpm2.SessionContext) (*tpm2.NVPublic, uint64, error) {
-	return createPcrPolicyCounterImpl(tpm, handle, updateKey, computeV3PcrPolicyCounterAuthPolicies, hmacSession)
-}
+// policy that permits TPM2_NV_Increment with a signed authorization policy.
+var ensurePcrPolicyCounter = func(tpm *tpm2.TPMContext, handle tpm2.Handle, updateKey *tpm2.Public, hmacSession tpm2.SessionContext) (public *tpm2.NVPublic, err error) {
+	nameAlg := tpm2.HashAlgorithmSHA256
 
-// createPcrPolicyCounterLegacy creates and initializes a NV counter that is associated with a sealed key object
-// and is used for implementing PCR policy revocation.
-//
-// The NV index will be created with attributes that allow anyone to read the index, and an authorization
-// policy that permits TPM2_NV_Increment with a signed authorization policy. The caller must ensure that the
-// updateKey argument is a valid public key.
-func createPcrPolicyCounterLegacy(tpm *tpm2.TPMContext, handle tpm2.Handle, updateKey *tpm2.Public, hmacSession tpm2.SessionContext) (*tpm2.NVPublic, uint64, error) {
-	return createPcrPolicyCounterImpl(tpm, handle, updateKey, computeV2PcrPolicyCounterAuthPolicies, hmacSession)
+	authPolicies := computeV3PcrPolicyCounterAuthPolicies(nameAlg, updateKey.Name())
+
+	trial := util.ComputeAuthPolicy(nameAlg)
+	trial.PolicyOR(authPolicies)
+
+	public = &tpm2.NVPublic{
+		Index:      handle,
+		NameAlg:    tpm2.HashAlgorithmSHA256,
+		Attrs:      tpm2.NVTypeCounter.WithAttrs(tpm2.AttrNVPolicyWrite | tpm2.AttrNVAuthRead | tpm2.AttrNVPolicyRead | tpm2.AttrNVNoDA),
+		AuthPolicy: trial.GetDigest(),
+		Size:       8}
+
+	index, err := tpm.CreateResourceContextFromTPM(handle, hmacSession)
+	switch {
+	case tpm2.IsResourceUnavailableError(err, handle):
+		// ok, need to create
+		index, err = tpm.NVDefineSpace(tpm.OwnerHandleContext(), nil, public, hmacSession)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if err == nil {
+				return
+			}
+			tpm.NVUndefineSpace(tpm.OwnerHandleContext(), index, hmacSession)
```

Comment:
this bit of undo logic does not seem tested

Reply:
This code was copied from `createPcrPolicyCounter` where this path was also untested. It's a bit tricky to test this case without mocking the calls to go-tpm2 and there's a bunch of other paths that depend on TPM errors that aren't tested.

I think it's worth addressing these in follow-ups, but by:
- adding an API to github.com/canonical/go-tpm2/mssim to put the simulator into failure mode.
- adding an API to github.com/canonical/go-tpm2/testutil to call the new API after a specific number of commands are executed successfully.
- using the new APIs here to trigger failures at specific points.